pgjdbc/src/main/java/org/postgresql/core/AuthenticationPluginManager.java

```diff
@@ -37,12 +37,16 @@ private AuthenticationPluginManager() {
    * @return The password to use for authentication or null if none is available
    * @throws PSQLException Throws a PSQLException if the plugin class cannot be instantiated
    */
-  public static @Nullable String getPassword(AuthenticationRequestType type, Properties info) throws PSQLException {
+  public static char @Nullable [] getPassword(AuthenticationRequestType type, Properties info) throws PSQLException {
```

Comment:
What do you think of adding the following to the javadoc?

```
The callers should erase the contents of `char[] password` with 0 chars like in
{@code Arrays.fill(password, (char) 0)} after the password is used for security reasons.
```

Reply:
Yes that's a good idea. I'm going to add that to the javadoc for the interface as that's the only part that a user would interact with. And it will be the user's responsibility to clear the array as they have ownership of it.

Reply:
Which “user” are you talking about? The caller of this api (eg. the driver itself as it needs password to connect) or the user of the driver who is implementing class to provides password dynamically?

It seems to me we should define the api to always have to return a new ‘char[]’ and make the driver responsible for clearing once value no longer needed.

Reply:
@vlsi @bokken Yes the more I'm thinking about this, the hairier it'd get to have the caller clear the `char[]` array as it's provided to us in the plugin interface, but the end user cannot clear it until after the connection is established which might be handled somewhere else entirely. I'm going to change the code to wipe it out after use and make a note in the Javadoc for the interface that any value provided will be destroyed after use. So the caller will be responsible for providing a *new* char[] array each time the same method is called.

Reply:
> Which “user” are you talking about? 

Anyone who calls `char @Nullable [] getPassword(` method.
Mainly I expect that it is the driver who will call the method.

However, the plugin implementation should understand how the value would be used,  so it still makes sense to clarify the intention:
1. The value could be erased after it is returned in `getPassword`. For instance, if the plugin requests the password from some sort of token server, the plugin should erase all the other copies of the password (e.g. clear buffers, etc, etc)
2. The caller of `getPassword` (==driver in 99% of the cases) would erase `char[]` after use

That ensures no passwords keep floating around after use.